### PR TITLE
Change maxReposToSearch default to be unlimited

### PR DESCRIPTION
**Pull request by nicksnyder**
I received a report from a user on Twitter that they were surprised their private Sourcegraph instance was limited to 50 repositories by default.
https://twitter.com/IvanZaQGrund/status/1114076394093457408

I took at look at the code and found a lot of inconsistencies. We prefill our default site config with a limit of 50, we document that the default is 500, and if the user specifies -1 it behaves as "unlimited". This seems unnecessarily complicated and confusing.

This PR changes the default behavior in all circumstances (unconfigured, 0, -1) to be "unlimited".

### CODE REVIEW

File: pkg/conf/confdefaults/confdefaults.go [hunk 1]
```diff
@@ -44,8 +44,7 @@ var DockerContainer = conftypes.RawUnified{
 	]
 }`,
 	Site: `{
-	"disablePublicRepoRedirects": true,
-	"maxReposToSearch": 50,
+	"disablePublicRepoRedirects": true
```
Comment (slimsag) [inline pkg/conf/confdefaults/confdefaults.go]: Setting this to unlimited in sourcegraph/server deployments is dangerous until we have indexing on by default there.

For example, a customer using server successfully today with 3k repos and users having to specify `repo:` would see their instance explode if we shipped this to them.

**Comment by slimsag** [comment]
I am fine with this for all deployments except server, because we need https://github.com/sourcegraph/sourcegraph/issues/2176 before doing that.

File: pkg/conf/confdefaults/confdefaults.go (hunk 1 above)
Comment (ijt) [inline pkg/conf/confdefaults/confdefaults.go]: Why is this change part of this PR?

Comment (nicksnyder) [reply pkg/conf/confdefaults/confdefaults.go]: What do you think is a reasonable default in this case? 50, 500?

Comment (nicksnyder) [reply pkg/conf/confdefaults/confdefaults.go]: Whatever we decide the most reasonable default is, we should make it the default without requiring an explicit configuration.

Comment (nicksnyder) [reply pkg/conf/confdefaults/confdefaults.go]: Not changing this value would mean we wouldn't not solve the issue reported by the user on Twitter.

Comment (ijt) [reply pkg/conf/confdefaults/confdefaults.go]: Pardon me. I misread this. For some reason I got an optical illusion telling me this change was setting `disablePublicRepoRedirects` to `true`, but it's removing a setting of `maxReposToSearch`. Have to read more carefully!

Comment (slimsag) [reply pkg/conf/confdefaults/confdefaults.go]: Yes, 50. That is 50 concurrent `git grep`s, some of which could be quite hefty depending on repository size.

**Comment by slimsag** [comment]
@nicksnyder I definitely see how it was confusing / bad code, but the current status quo is:

- 50 maxReposToSearch on sourcegraph/server instances
- 500 maxReposToSearch on cluster instances

It should be changed to:

- 50 maxReposToSearch on sourcegraph/server instances
- unlimited maxReposToSearch on cluster instances

**Comment by nicksnyder** [comment]
Keeping sourcegraph/server limited to 50 by default would not address the concern in the Tweet. Users like this might incorrectly assume that Sourcegraph can't search all of their code when it totally can (given enough hardware).

I still think the right thing to do is to remove this arbitrary limit by default and for the cases where someone does overload their instance let whatever happens happen. Either they will realize that they need to add more hardware, or they will file an issue and we can help them from there. As it is now, I am worried we have users silently drop off disappointed in a completely unnecessary way (i.e. the instance they configured is totally capable of searching across all repositories).

My mental model is that 100% of users want "search all of my code". Divide this into three groups given the current behavior (excluding this PR):
1. Users with fewer than 50 repos. sourcegraph/server "just works". This PR doesn't impact these users so no need to discuss this case further.
1. Users with more than 50 repos and their instance can handle it. One of the following cases happens:
    1. They don't discover `maxReposToSearch`, and their usage of Sourcegraph stops or doesn't grow because it isn't as useful to not be able to search across all code.
    1. They discover `maxReposToSearch`, set it to unlimited, and everything works (as what happen to the person on Twitter).
1. Users with more than 50 repos and their instance can't handle it.
    1. They don't discover `maxReposToSearch`, and their usage of Sourcegraph stops or doesn't grow because it isn't as useful to not be able to search across all code.
    1. They discover `maxReposToSearch`, set it to unlimited, and their instance misbehaves in some way (e.g. slow results, OOM).
        1. They add more hardware until it works.
        1. They file an issue and ask for help.
        1. They tune `maxReposToSearch` to not be unlimited, but greater than 50.
        1. They give up and stop using Sourcegraph.

With my PR:
- (2) reduces to "it just works" like (1)
- (3) reduces to "their instance misbehaves in some way (e.g. slow results, OOM)." and the same followup options apply.

Both of the above seem like good simplifications/optimizations from a user perspective to me (especially 2).

**Comment by sqs** [comment]
I agree with @nicksnyder. While there is no perfect outcome that we can reach with a limited amount of work, what he describes is better than what we have today.

**Comment by slimsag** [comment]
> My mental model is that 100% of users want "search all of my code".

Your assumption, also reflected in your third point, is that nobody will use or find value in Sourcegraph to search just a single repository or small subset of some >50 set. But is that true?

This would mean that **every user of server to date** has had <50 repositories OR has discovered `maxReposToSearch` and set it to unlimited.

If you are certain that is the case, then I am fine with this change.

I however fear that we have a non-trivial amount of users in this category, and when we ship them this version of Sourcegraph they will begin to encounter slow results, OOMs, and other resource constraint issues.

Two possible ways to alleviate my fear:

1. On old instances, still default to a max of 50, so that upgrading is not a regression for them. Don't know how we would do this from a technical POV.
2. Make it **extremely** clear that Sourcegraph server now searches over all repositories by default _and that this can consume multiples of resources that the prior version required._ A changelog entry or minor blog post line would not be enough as it could be easily missed.

**Comment by sqs** [comment]
> On old instances, still default to a max of 50, so that upgrading is not a regression for them. Don't know how we would do this from a technical POV.

The default would still be set to 50 for old instances because the default config is stored in the DB at the time the site was first run, right? So this is what would happen (without any more changes beyond merging this PR).

> Make it extremely clear that Sourcegraph server now searches over all repositories by default and that this can consume multiples of resources that the prior version required. A changelog entry or minor blog post line would not be enough as it could be easily missed.

Two things help here:

- The existing timeout error message has helpful guidance here already: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f6a6de81d2a24bb9cc14e4fae872b33bb6b35f13/-/blob/web/src/search/results/SearchResultsList.tsx#L402-412 (note there is likely a logic bug, just tagged @attfarhan here).
- The new repo syncing behavior in #2025 helps here because people are a lot less likely to add a ton of repositories that they don't actually care about.

And what about the following other changes (I agree a changelog notice is likely to be missed):

- Set the default timeout to 15s on sourcegraph/server instances, not 60s. This is because someone using sourcegraph/server is less likely to be interested in having a long-running search across a ton of code, and shortening the timeout will show them the useful info notice above sooner.
- Add a global alert banner for site admins if there are more than ~50 repositories on the instance referring them to our scaling docs and noting that sourcegraph/server scalability is highly dependent on the hardware.

**Comment by nicksnyder** [comment]
> On old instances, still default to a max of 50, so that upgrading is not a regression for them. Don't know how we would do this from a technical POV.

As Quinn said, I don't think we need to do anything here since this config would be stored in the DB for those customers and behavior wouldn't change. I think a changelog would be sufficient.

> Your assumption, also reflected in your third point, is that nobody will use or find value in Sourcegraph to search just a single repository or small subset of some >50 set.

I think those are valuable use cases, but neither of those use cases are made worse by changing the default of maxReposToSearch to unlimited. To search a single repo, you have to specify a single repo filter. This works regardless of how maxReposToSearch is configured. Similarly if there is an instance with >50 repos but is configured with the default of 50 maxReposToSearch, then all current search queries will need to explicitly specify a repo filter that matches <50. Those queries would work the same as before.

My instinct is that any customer who is in the state of (3)(i) would prefer to transition to a state of (3)(ii)(b) if they knew that was possible.

> This would mean that every active user of server to date has had <50 repositories OR has discovered maxReposToSearch and set it to unlimited.

OR they haven't experienced the true value of Sourcegraph because they can't search all their code and don't know what they are missing. This is a degenerate case in my opinion. If they knew what they were missing, the would _want_ to set it to unlimited and allocate more resources to their instance.

> Set the default timeout to 15s on sourcegraph/server instances, not 60s. This is because someone using sourcegraph/server is less likely to be interested in having a long-running search across a ton of code, and shortening the timeout will show them the useful info notice above sooner.

This seems like a marginal change/win.

> Add a global alert banner for site admins if there are more than ~50 repositories on the instance referring them to our scaling docs and noting that sourcegraph/server scalability is highly dependent on the hardware.

This makes a lot more sense to me as long as it is dismissible. Even better would be if the Sourcegraph instance could detect if it didn't have enough resources available (instead of a hard coded repo count), but that would be a more sophisticated change.

In any case, I think those improvements are out of scope for this PR.

**Comment by nicksnyder** [comment]
I am going to merge this after CI is green again.

https://github.com/sourcegraph/sourcegraph/pull/3374 includes a new message about performance when more than 100 repos are added to a single node instance.

**Pull request merged**